### PR TITLE
Add joined(separator:)

### LINKED ISSUE #984: func joined(separator:) for NSAttributedStrings
🚀
**Is your feature request related to a problem? Please describe.**
In my projects I use a lot of `NSAttributedString`s, but did not find a convenient way how to create an `NSAttributedString` instance with different attributes for each range.

**Describe the solution you'd like**
I would like to add these functions:
```
public extension Array where Element: NSAttributedString {
    // https://stackoverflow.com/questions/32830519/is-there-joinwithseparator-for-attributed-strings
    func joined(separator: NSAttributedString) -> NSAttributedString {
        var isFirst = true
        return reduce(into: NSMutableAttributedString()) { result, element in
            if isFirst {
                isFirst = false
            } else {
                result.append(separator)
            }
            result.append(element)
        }
    }

    func joined(separator: String) -> NSAttributedString {
        joined(separator: NSAttributedString(string: separator))
    }
}
```

- [x] I've read, understood, and done my best to follow the [Contributing guidelines](https://github.com/SwifterSwift/SwifterSwift/tree/master/CONTRIBUTING.md) before opening this issue.

### PR DESCRIPTION
**Pull request by RomanPodymov**
Hello.
Thank you for SwifterSwift.
This pull request fixes #984.

## Checklist

- [x] I checked the [**Contributing Guidelines**](https://github.com/SwifterSwift/SwifterSwift/blob/master/CONTRIBUTING.md) before creating this request.
- [x] New extensions are written in Swift 5.0.
- [x] New extensions support iOS 10.0+ / tvOS 9.0+ / macOS 10.10+ / watchOS 2.0+, or use `@available` if not.
- [x] I have added tests for new extensions, and they passed.
- [x] All extensions have a **clear** comments explaining their functionality, all parameters and return type in English.
- [x] All extensions are declared as **public**.
- [x] I have added a [changelog](https://github.com/SwifterSwift/SwifterSwift/blob/master/CHANGELOG_GUIDELINES.md) entry describing my changes.

### CODE REVIEW

File: Sources/SwifterSwift/Foundation/NSAttributedStringExtensions.swift [hunk 1]
```diff
@@ -180,4 +180,26 @@ public extension NSAttributedString {
...
+    func joined(separator: NSAttributedString) -> NSAttributedString {
+        var isFirst = true
+        return reduce(into: NSMutableAttributedString()) { result, element in
+            if isFirst {
+                isFirst = false
+            } else {
+                result.append(separator)
+            }
+            result.append(element)
+        }
```
Comment (LucianoPAlmeida) [inline Sources/SwifterSwift/Foundation/NSAttributedStringExtensions.swift:198]: Here is an implementation suggestion to avoid having to check `isFirst` every iteration
```suggestion
        guard let firstElt = self.first else { return NSMutableAttributedString() }
        var mStr = NSMutableAttributedString(firstElt)
        return dropFirst().reduce(into: mStr) { result, element in
            result.append(separator)
            result.append(element)
        }
```
Maybe not even need a reduce into, a for-loop would be fine :)

**Review by LucianoPAlmeida — COMMENTED** [review]
Thanks for the contribution @RomanPodymov :)
I've left a nit suggestion inline

File: Sources/SwifterSwift/Foundation/NSAttributedStringExtensions.swift (hunk 1 above)
Comment (RomanPodymov) [reply Sources/SwifterSwift/Foundation/NSAttributedStringExtensions.swift]: Hello @LucianoPAlmeida 
Thank you for your suggestion, let's use it.

**Review by LucianoPAlmeida — DISMISSED** [review]
Looks fine to me :)

File: Sources/SwifterSwift/Foundation/NSAttributedStringExtensions.swift [hunk 2]
```diff
@@ -180,4 +180,22 @@ public extension NSAttributedString {
...
 }
 
+public extension Array where Element: NSAttributedString {
+    /// SwifterSwift: Returns a new `NSAttributedString` by concatenating the elements of the sequence, adding the given separator between each element.
+    ///
+    /// - Parameter separator: An `NSAttributedString` to add between the elements of the sequence.
+    /// - Returns: NSAttributedString with applied attributes.
+    // https://stackoverflow.com/questions/32830519/is-there-joinwithseparator-for-attributed-strings
+    func joined(separator: NSAttributedString) -> NSAttributedString {
+        guard let firstElt = first else { return NSMutableAttributedString() }
```
Comment (LucianoPAlmeida) [inline Sources/SwifterSwift/Foundation/NSAttributedStringExtensions.swift:190]: This seems to be failing on linux, but I think this should fix it :)
```suggestion
        guard let firstElt = first else { return NSMutableAttributedString(string: "") }
```

File: Sources/SwifterSwift/Foundation/NSAttributedStringExtensions.swift [hunk 3]
```diff
@@ -180,4 +180,22 @@ public extension NSAttributedString {
...
+        return dropFirst().reduce(into: NSMutableAttributedString(attributedString: firstElt)) { result, element in
+            result.append(separator)
+            result.append(element)
+        }
+    }
+
+    func joined(separator: String) -> NSAttributedString {
+        joined(separator: NSAttributedString(string: separator))
```
Comment (guykogus) [inline Sources/SwifterSwift/Foundation/NSAttributedStringExtensions.swift:198]: One of the rules of the project is that functions can't depend on one another, so that they can be copy-pasted individually. It leads to code-copying, but it makes life easier as it's often how the project is used. So you'll need to copy the code from the other `joined` function.

File: Sources/SwifterSwift/Foundation/NSAttributedStringExtensions.swift (hunk 2 above)
Comment (guykogus) [inline Sources/SwifterSwift/Foundation/NSAttributedStringExtensions.swift:190]: Avoid using shorthands like "elt" for element. It can make code difficult to read.
```suggestion
        guard let firstElement = first else { return NSMutableAttributedString() }
```

File: Tests/FoundationTests/NSAttributedStringExtensionsTests.swift [hunk 4]
```diff
@@ -319,6 +319,47 @@ final class NSAttributedStringExtensionsTests: XCTestCase {
...
+                .expansion: NSNumber(value: 3),
+                .obliqueness: NSNumber(value: 4)
+            ]
+        )
+        let string3 = NSAttributedString(string: "World", attributes: [:])
+        let expectation = NSMutableAttributedString(string: "Hello World", attributes: [:])
+        expectation.addAttribute(
+            .strokeWidth,
+            value: NSNumber(value: 1),
+            range: .init(location: 0, length: "Hello".count)
```
Comment (guykogus) [inline Tests/FoundationTests/NSAttributedStringExtensionsTests.swift:343]: We don't use `.init` in the project. Better to explicitly call `Range(...)`

File: Tests/FoundationTests/NSAttributedStringExtensionsTests.swift [hunk 5]
```diff
@@ -319,6 +319,47 @@ final class NSAttributedStringExtensionsTests: XCTestCase {
...
+            value: NSNumber(value: 3),
+            range: .init(location: "Hello".count, length: " ".count)
+        )
+        expectation.addAttribute(
+            .obliqueness,
+            value: NSNumber(value: 4),
+            range: .init(location: "Hello".count, length: " ".count)
+        )
+        XCTAssertEqual([string1, string2, string3].joined(separator: ""), expectation)
+        XCTAssertEqual([].joined(separator: NSAttributedString(string: "Hello")), NSAttributedString(string: ""))
```
Comment (guykogus) [inline Tests/FoundationTests/NSAttributedStringExtensionsTests.swift:361]: Can you add tests for non-empty string separator, and an empty attributed string separator?

**Review by guykogus — CHANGES_REQUESTED** [review]

**Review by LucianoPAlmeida — APPROVED** [review]

File: Tests/FoundationTests/NSAttributedStringExtensionsTests.swift [hunk 6]
```diff
@@ -319,6 +319,148 @@ final class NSAttributedStringExtensionsTests: XCTestCase {
         XCTAssertEqual(string1.string, "Test Appending")
         #endif
     }
+
+    // MARK: - func joined(separator:)
+
+    func testJoinedWithEmptySeparator() {
+        #if canImport(AppKit) || canImport(UIKit)
```
Comment (guykogus) [inline Tests/FoundationTests/NSAttributedStringExtensionsTests.swift:326]: Cover the test functions themselves, not the content.

File: Tests/FoundationTests/NSAttributedStringExtensionsTests.swift [hunk 7]
```diff
@@ -319,6 +319,148 @@ final class NSAttributedStringExtensionsTests: XCTestCase {
...
+            value: NSNumber(value: 4),
+            range: NSRange(location: "Hello".count, length: " ".count)
+        )
+        XCTAssertEqual([string1, string2, string3].joined(separator: ""), expectation)
+        #endif
+    }
+
+    func testJoinedWithEmptyAttributedSeparator() {
+        #if canImport(AppKit) || canImport(UIKit)
+        let string1 = NSAttributedString(
```
Comment (guykogus) [inline Tests/FoundationTests/NSAttributedStringExtensionsTests.swift:369]: In the tests you can extract functions for reuse. The same rules don't apply here because this isn't what users are importing/copying :)

**Review by guykogus — CHANGES_REQUESTED** [review]

**Comment by LucianoPAlmeida** [comment]
@RomanPodymov Any updates on this? :)

**Comment by RomanPodymov** [comment]
> @RomanPodymov Any updates on this? :)

Hello @LucianoPAlmeida 
Sorry for the delay in responding, please check the latest version.

File: Tests/FoundationTests/NSAttributedStringExtensionsTests.swift [hunk 8]
```diff
@@ -319,6 +319,114 @@ final class NSAttributedStringExtensionsTests: XCTestCase {
...
+            separatorAttrbiutes,
+            range: NSRange(location: (firstStringToJoin + separator + secondStringToJoin).count, length: separator.count)
+        )
+
+        return expectation
+    }
+    #endif
+    
+    func testJoinedWithEmptySeparator() {
+        #if canImport(AppKit) || canImport(UIKit)
```
Comment (LucianoPAlmeida) [inline Tests/FoundationTests/NSAttributedStringExtensionsTests.swift:383]: Could this be outside the method as well? Pehaps everything should be enclosed on the block above

**Comment by LucianoPAlmeida** [comment]
> > @RomanPodymov Any updates on this? :)
> 
> Hello @LucianoPAlmeida Sorry for the delay in responding, please check the latest version.

No worries! Everything looks ok, just one minor question

**Review by LucianoPAlmeida — APPROVED** [review]
Thanks
